### DISCUSSION TITLE
Move client email templates to Firestore

### DISCUSSION THREAD

**Discussion by we-ai** [episphere/connect#812] [General]
Currently we have 2 types of email reminders sent to participants: scheduled reminders (sent at 3 PM or 8 PM each day) and client side reminders (sent immediately when conditions met). They're store in different places and use different triggers.

The templates of scheduled email reminders are stored in Firestore. They're triggered by time (3 PM or 8 PM each day based on settings in schema).

We have 1 client side reminder email template (category: "Biospecimen Survey Reminder", attempt: "1st contact") coded in our biospecimen repo. With progress of Connect project, we're adding more client side reminders, and each reminder will be coded in our repo if we follow current practice creating client side reminders. Putting email templates in our code repo means that we have to update our code when a template needs updates, which is less efficient compared to updating scheduled email reminder templates. A better way is to put client side email templates in Firestore.

Last week, Davin and I briefly discussed the feasibility of storing client side email templates in Firestore. After checking related code in biospecimen repo, I believe it can be done. In November release of 2023, we're adding 3 more client side email templates. We can store these new templates in Firesotre, and move existing client side email template to Firestore later (in next release). Whether client side email templates should be stored in the same collection (`notificationSpecifications`) of scheduled reminders or in a new collection (eg `clientSideEmails`) is open to discussion. I prefer a new collection to store client side email templates since overall schemas of these 2 types of reminders have quite different data structures.

**Comment by brotzmanmj** [comment]
@we-ai  thanks for the explanation. How does this impact the November timeline for programming and testing of the three new client side emails for home collections?

**Reply by we-ai** [reply]
I believe everything (timeline, testing, etc.) should proceed as planned. @brotzmanmj

**Comment by anthonypetersen** [comment]
Initial Thoughts... Client side emails and server side (scheduled) emails share a lot of fields and it feels unnecessary to create a new collection just for client side emails. We could easily add a field on this collection `type` that we could integrate into how we query these specifications from our code.

**Reply by we-ai** [reply]
The key can be `scheduleType`, with possible values of `scheduled` (for scheduled reminder; default value) or `immediate` (for client side reminder).
"scheduleType": "scheduled" | "immediate"

**Reply by anthonypetersen** [reply]
I like those, very clear

**Reply by we-ai** [reply]
Maybe use `sendType` as a key? It's better than `scheduleType`.
"sendType": "scheduled" | "instant"

**Reply by anthonypetersen** [reply]
Either of those options work for me, both are clear.

**Comment by we-ai** [comment]
Updates in biospecimen ([PR#711](https://github.com/episphere/biospecimen/pull/711)) and ConnectFaaS ([PR#603](https://github.com/episphere/connectFaas/pull/603)) completed the implementation in June (2024) release.
Discussed requirements were met.
